cluster/ci/origin/pipelines/build/Jenkinsfile

```diff
@@ -46,107 +49,183 @@ pipeline {
-                "RELEASE_SRC_URL=${env.RELEASE_SRC_URL ?: 'https://github.com/openshift/release.git'}",
-                "PUSH_SECRET_NAME=${this._pushSecret}",
-                "BUILD_NAME=${this._buildName}",
-                "REPO_OWNER=${params.REPO_OWNER}",
-                "REPO_NAME=${params.REPO_NAME}",
-                "PULL_REFS=${params.PULL_REFS}",
+              return createTemplate(this, "${templateBase}/build-rpms.yaml",
+                "RELEASE_SRC_REF=${params.RELEASE_SRC_REF}",
+                "RELEASE_SRC_URL=${params.RELEASE_SRC_URL}", 
+                "IMAGE_SECRET_NAME=${this._pushSecret}",
+                "BASE_IMAGE_REF=build-base:latest",
+                "RELEASE_SRC_CONTEXT=cluster/ci/origin/images/clean-rpms",
                 "PULL_NUMBER=${params.PULL_NUMBER}",
-                "BUILD_NUMBER=${env.BUILD_NUMBER}")
-              return launchedBuilds["clone-source"]
+                "BUILD_NUMBER=${env.BUILD_NUMBER}",
+                "BUILD_NAME=${this._buildName}",
+                "JOB_ID=${this._jobId}")
             })
           } else {
-            echo "Skipping source build for ${this._buildName}. Another build is already running: ${activeBuilds[0].metadata.name}"
+            runAndWaitForBuild(this, "", {
+              return createTemplate(this, "${templateBase}/build-rpms.yaml",
+                "RELEASE_SRC_REF=${params.RELEASE_SRC_REF}",
+                "RELEASE_SRC_URL=${params.RELEASE_SRC_URL}", 
+                "IMAGE_SECRET_NAME=${this._pushSecret}",
+                "BASE_IMAGE_REF=${this._buildName}:binaries",
+                "PULL_NUMBER=${params.PULL_NUMBER}",
+                "BUILD_NUMBER=${env.BUILD_NUMBER}",
+                "BUILD_NAME=${this._buildName}",
+                "JOB_ID=${this._jobId}")
+            })
           }
         }
       }
     }
-    stage("Compile") {
+    stage("Serve RPMs") {
+      when {
+        not { expression { return serviceExists(this, "rpm-repo-${this._buildName}") } }
+      }
+      steps {
+        script {
+          def imageStreamRepo
+          openshift.withCluster() {
+            imageStreamRepo = openshift.selector("is/${this._buildName}").object().status.dockerImageRepository
+          }
+          applyTemplate(this, "${templateBase}/serve-rpms.yaml", 
+            "RPM_IMAGE_REF=${imageStreamRepo}:rpms",
+            "BUILD_NAME=${this._buildName}")
+          waitForDeployment(this, "rpm-repo-${this._buildName}")
+        }
+      }
+    }
+    stage("Build Repository Base") {
+      when {
+        not { expression { return imageStreamTagExists(this, "${this._buildName}", "repo-base") } }
+      }
+      steps {
+        script {
+          def routeHost
+          openshift.withCluster() {
+            routeHost = openshift.selector("route/rpm-repo-${this._buildName}").object().spec.host
+          }
+          runAndWaitForBuild(this, "", {
+            return createTemplate(this, "${templateBase}/repo-base.yaml",
```

Comment:
why do we do this after building and serving the RPMs?

Reply:
It takes the route as an argument (rpm repo url). We only know the hostname of the route after creating the previous deployment.

Reply:
Why do we need to serve the RPMs over a route vs a clusterip service?

Reply:
because the cluster will be provisioned elsewhere in GCE. The vm's being provisioned will need to access the rpm repository we serve here.

Reply:
ack